stripecardscan/src/main/java/com/stripe/android/stripecardscan/cardscan/CardScanActivity.kt

```diff
@@ -0,0 +1,571 @@
+            false,
+        ) -> showPermissionDeniedDialog()
+        else -> requestCameraPermission()
+    }
+
+    /**
+     * Once the camera stream is available, start processing images.
+     */
+    private fun onCameraStreamAvailable(cameraStream: Flow<CameraPreviewImage<Bitmap>>) {
+        scanFlow.startFlow(
+            context = this,
+            imageStream = cameraStream,
+            viewFinder = viewBinding.viewFinderWindow.asRect(),
+            lifecycleOwner = this,
+            coroutineScope = this
+        )
+    }
+
+    /**
+     * Show a dialog explaining that the camera is not available.
+     */
+    private fun showCameraNotSupportedDialog() {
+        AlertDialog.Builder(this)
+            .setTitle(R.string.stripe_error_camera_title)
+            .setMessage(R.string.stripe_error_camera_unsupported)
+            .setPositiveButton(R.string.stripe_error_camera_acknowledge_button) { _, _ ->
+                scanFailure()
+            }
+            .show()
+    }
+
+    /**
+     * Show an explanation dialog for why we are requesting camera permissions.
+     */
+    private fun showPermissionRationaleDialog() {
+        val builder = AlertDialog.Builder(this)
+        builder.setMessage(R.string.stripe_camera_permission_denied_message)
+            .setPositiveButton(R.string.stripe_camera_permission_denied_ok) { _, _ ->
+                requestCameraPermission()
+            }
+        builder.show()
+        storage.storeValue(PERMISSION_RATIONALE_SHOWN, true)
+    }
+
+    /**
+     * Show an explanation dialog for why we are requesting camera permissions when the permission
+     * has been permanently denied.
+     */
+    private fun showPermissionDeniedDialog() {
+        val builder = AlertDialog.Builder(this)
+        builder.setMessage(R.string.stripe_camera_permission_denied_message)
+            .setPositiveButton(R.string.stripe_camera_permission_denied_ok) { _, _ ->
+                storage.storeValue(PERMISSION_RATIONALE_SHOWN, false)
+                openAppSettings()
+            }
+            .setNegativeButton(R.string.stripe_camera_permission_denied_cancel) { _, _ ->
+                userDeniedCameraPermission()
+            }
+        builder.show()
+    }
+
+    /**
+     * Request permission to use the camera.
+     */
+    private fun requestCameraPermission() {
+        ActivityCompat.requestPermissions(
+            this,
+            arrayOf(Manifest.permission.CAMERA),
+            PERMISSION_REQUEST_CODE,
+        )
+    }
+
+    /**
+     * Open the settings for this app
+     */
+    private fun openAppSettings() {
+        val intent = Intent()
+            .setAction(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
+            .setData(Uri.fromParts("package", getAppPackageName(this), null))
+        startActivity(intent)
+    }
+
+    /**
+     * Cancel scanning due to a camera error.
+     */
+    private fun scanFailure(cause: Throwable? = null) {
```

Comment:
is it possible to share the scan closing logic (scanFailure, userClosedScanner, userDeniedCameraPermission) with the CardImageVerificationActivity through the shared super class?